crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -746,6 +750,49 @@ impl AuthorityEpochTables {
         batch.write()?;
         Ok(())
     }
+
+    pub fn load_initial_object_debts(
+        &self,
+        current_round: Round,
+        for_randomness: bool,
+        per_commit_budget: u64,
+        transactions: &[VerifiedSequencedConsensusTransaction],
+    ) -> SuiResult<impl IntoIterator<Item = (ObjectID, u64)>> {
+        let table = if for_randomness {
+            &self.congestion_control_randomness_object_debts
+        } else {
+            &self.congestion_control_object_debts
+        };
+        let shared_input_object_ids: BTreeSet<_> = transactions
+            .iter()
+            .filter_map(|tx| {
+                if let SequencedConsensusTransactionKind::External(ConsensusTransaction {
+                    kind: ConsensusTransactionKind::CertifiedTransaction(tx),
+                    ..
+                }) = &tx.0.transaction
+                {
+                    Some(tx.shared_input_objects().map(|obj| obj.id))
+                } else {
+                    None
+                }
+            })
+            .flatten()
+            .collect();
+        Ok(table
+            .multi_get(shared_input_object_ids.iter())?
+            .into_iter()
+            .flatten()
+            .zip(shared_input_object_ids)
+            .map(move |((round, debt), object_id)| {
+                (
+                    object_id,
+                    // Stored debts already account for the budget of the round in which
+                    // they were accumulated. Application of budget from future rounds to
+                    // the debt is handled here.
+                    debt.saturating_sub(per_commit_budget * (current_round - round - 1)),
```

Comment:
Question: reading more closely and IIUC, `current_round > round + 1` can only happen when the object removed all the debts in round `round`. Why not just remove it when objects' debt drop to 0? If we can do so, do we still need to track round in the table?

Reply:
Am I reading the code correct that `accumulated_debts` will update all objects' current debt, including ones that are not used in the current commit, which means that r2 will also update o1's debt?

Reply:
Oh, we load objects that are only used in the current commit. That makes sense now.